### PR TITLE
QOS fieldvalue reference ABNF format to string

### PR DESCRIPTION
**Pull request by AshokDaparthi**
Depends on https://github.com/Azure/sonic-swss/pull/1754

Qos tables in config db and app db used ABNF format i.e "[TABLE_NAME|name] to refer fieldvalue other qos tables.

Example:
Config DB:
"Ethernet92|3": {
"scheduler": "[SCHEDULER|scheduler.1]",
"wred_profile": "[WRED_PROFILE|AZURE_LOSSLESS]"
},
"Ethernet0|0": {
"profile": "[BUFFER_PROFILE|ingress_lossy_profile]"
},
"Ethernet0": {
"dscp_to_tc_map": "[DSCP_TO_TC_MAP|AZURE]",
"pfc_enable": "3,4",
"pfc_to_queue_map": "[MAP_PFC_PRIORITY_TO_QUEUE|AZURE]",
"tc_to_pg_map": "[TC_TO_PRIORITY_GROUP_MAP|AZURE]",
"tc_to_queue_map": "[TC_TO_QUEUE_MAP|AZURE]"
},

AppDB:
"BUFFER_QUEUE_TABLE:Ethernet88:3-4": {
"profile": "[BUFFER_PROFILE_TABLE:egress_lossless_profile]"
},

1#This format is not consistent with other DB schema followed in sonic.
2# Added db_migrator.py case to  change from old format in config_db and appl_db  to new format. 
3#Modified the test case 

Dependent pull requests: 
https://github.com/Azure/sonic-buildimage/pull/7752  - To modify platfrom files 
https://github.com/Azure/sonic-buildimage/pull/7281 - Yang model 
https://github.com/Azure/sonic-swss/pull/1754    - swss change to remove ABNF format

### CODE REVIEW

**Review by stephenxs — CHANGES_REQUESTED** [review]
Hi @AshokDaparthi 
Some other parts need to be updated as well:
- The buffer configuration is stored in APPL_DB. The warm reboot will fail if the switch is upgrading from an image before this change to one after. So db migrator needs to handle APPL_DB as well.
- Suggest adding unit test cases for db migrator.
- The version defined at the beginning of [db_migrator.py](https://github.com/Azure/sonic-utilities/blob/master/scripts/db_migrator.py#L47) needs to be updated as well.
- Existing unit test cases need to be updated accordingly, like [mock_tables](https://github.com/Azure/sonic-utilities/tree/master/tests/mock_tables), and mock tables for [qos](https://github.com/Azure/sonic-utilities/tree/master/tests/qos_config_input), [filter fdb](https://github.com/Azure/sonic-utilities/tree/master/tests/filter_fdb_input), [counter poll](https://github.com/Azure/sonic-utilities/tree/master/tests/counterpoll_input), [SKU create tool](https://github.com/Azure/sonic-utilities/tree/master/tests/sku_create_input), etc... You might want to go over all the test cases.
 I believe the unit test is failing without these changes

Thanks

**Comment by AshokDaparthi** [comment]
@stephenxs -                                                                                                                                                                                                   Hi Stephan,                                                                                                                                                                                                          I have query regarding "The buffer configuration is stored in APPL_DB. The warm reboot will fail if the switch is upgrading from an image before this change to one after. So db migrator needs to handle APPL_DB as well."                                                                       I have not seen complete warm reboot design. Can you help me to understand, While warm/cold reboot db_migratory.py will trigger config_db update for qos right, which internally trigger swss to update APP_DB with new QOS field values right or during warmreboot swss will not process config_db updates?

**Comment by stephenxs** [comment]
> @stephenxs - Hi Stephan, I have query regarding "The buffer configuration is stored in APPL_DB. The warm reboot will fail if the switch is upgrading from an image before this change to one after. So db migrator needs to handle APPL_DB as well." I have not seen complete warm reboot design. Can you help me to understand, While warm/cold reboot db_migratory.py will trigger config_db update for qos right, which internally trigger swss to update APP_DB with new QOS field values right or during warmreboot swss will not process config_db updates?

The buffer tables are consumed from APPL_DB instead of CONFIG_DB. 
In a normal starting flow,
- the buffer manager generates table entries and pushes them into APPL_DB. 
- the orchagent consumes the entries in APPL_DB and pushes entries into ASIC_DB.

This is for both dynamic and traditional buffer models.
In a warm reboot flow,
- the `APPL_DB` is saved before warm reboot and restored after it, which means the entries consumed by the orchagent are restored instead of pushed by the manager daemons.
- the buffer manager generates table entries as well, but that's too late for warm reboot to meet its converging time.
- given the data format has been updated before and after warm reboot, we need to update the entries in APPL_DB as well

Hope this helps.

File: config/main.py [hunk 1]
```diff
@@ -3703,7 +3703,7 @@ def update_pg(ctx, interface_name, pg_map, override_profile, add = True):
             ctx.fail("Profile {} doesn't exist".format(override_profile))
         if not 'xoff' in profile_dict.keys() and 'size' in profile_dict.keys():
             ctx.fail("Profile {} doesn't exist or isn't a lossless profile".format(override_profile))
-        profile_full_name = "[BUFFER_PROFILE|{}]".format(override_profile)
+        profile_full_name = "{}".format(override_profile)
```
[the remark this answers is not in the document]
Comment (AshokDaparthi) [reply config/main.py]: Done

File: config/main.py [hunk 2]
```diff
@@ -4775,7 +4775,7 @@ def remove_profile(db, profile):
     config_db = db.cfgdb
     ctx = click.get_current_context()
 
-    full_profile_name = '[BUFFER_PROFILE|{}]'.format(profile)
+    full_profile_name = '{}'.format(profile)
```
[the remark this answers is not in the document]
Comment (AshokDaparthi) [reply config/main.py]: Done

File: scripts/db_migrator.py [hunk 3]
```diff
@@ -382,6 +384,69 @@ def migrate_config_db_port_table_for_auto_neg(self):
                 elif value['autoneg'] == '0':
                     self.configDB.set(self.configDB.CONFIG_DB, '{}|{}'.format(table_name, key), 'autoneg', 'off')
 
+    def migrate_qos_db_fieldval_reference_remove(self, table_list=None, db=None, db_num=None, db_delimeter=''):
```
[the remark this answers is not in the document]
Comment (AshokDaparthi) [reply scripts/db_migrator.py]: Done

File: scripts/db_migrator.py [hunk 4]
```diff
@@ -382,6 +384,69 @@ def migrate_config_db_port_table_for_auto_neg(self):
...
+            table_name, fields_list = pair
+            qos_table = db.get_table(table_name)
+            for key, value in qos_table.items():
+                if type(key) is tuple:
+                    db_key = table_name + db_delimeter + db_delimeter.join(key)
+                else:
+                    db_key = table_name + db_delimeter + key
+
+                for field in fields_list:
+                  if field in value:
```
[the remark this answers is not in the document]
Comment (AshokDaparthi) [reply scripts/db_migrator.py]: Done

File: scripts/db_migrator.py [hunk 5]
```diff
@@ -382,6 +384,69 @@ def migrate_config_db_port_table_for_auto_neg(self):
...
+                     fieldVal = value.get(field)
+                     if not fieldVal or fieldVal == "NULL":
+                        continue
+                     newFiledVal = ""
+                     # Check for ABNF format presence and convert ABNF to string
+                     if "[" in fieldVal and db_delimeter in fieldVal and "]" in fieldVal:
+                        log.log_info("Found ABNF format field value in table {} key {} field {} val {}".format(table_name, db_key, field, fieldVal))
+                        value_list = fieldVal.split(",")
+                        for item in value_list:
+                          if "[" not in item or db_delimeter not in item or "]" not in item:
```
[the remark this answers is not in the document]
Comment (AshokDaparthi) [reply scripts/db_migrator.py]: Done

File: scripts/db_migrator.py [hunk 6]
```diff
@@ -537,6 +602,7 @@ def version_2_0_1(self):
             for name, data in portchannel_table.items():
                 data['lacp_key'] = 'auto'
                 self.configDB.set_entry('PORTCHANNEL', name, data)
+        self.migrate_qos_fieldval_reference_format()
```
[the remark this answers is not in the document]
Comment (AshokDaparthi) [reply scripts/db_migrator.py:605]: When i am doing it 2_0_2 is new one, i added new version 2_0_3

**Review by stephenxs — APPROVED** [review]

**Comment by AshokDaparthi** [comment]
@stephenxs , @neethajohn - Could you please re approve.                                                                                                                                  Plan is to 
1# Merge sonic-utilities 
2# Modify SHA                                                                                                                                                                                                   3# Then re trigger sonic-swss Checks and merge it                                                                                                                                     4# Merge sonic-buildimage

**Comment by AshokDaparthi** [comment]
> 
> 
> > @stephenxs , @neethajohn - Could you please re approve. Plan is to
> > 1# Merge sonic-utilities
> > 2# Modify SHA 3# Then re trigger sonic-swss Checks and merge it 4# Merge sonic-buildimage
> 
> Hi @AshokDaparthi
> Two questions.
> 
>     1. What's the difference between this one and the commit we approved? Is it just rebased to the latest master commit?
> 
>     2. To what branch will this feature be merged?
>        I was wondering whether we will have a dedicated sonic-mgmt branch for testing sonic switch without this feature. Check [this PR](https://github.com/Azure/sonic-mgmt/pull/3824#issuecomment-906797660) for details
>        @neethajohn FYI.
> 
> 
> Thanks



> 
> 
> > @stephenxs , @neethajohn - Could you please re approve. Plan is to
> > 1# Merge sonic-utilities
> > 2# Modify SHA 3# Then re trigger sonic-swss Checks and merge it 4# Merge sonic-buildimage
> 
> Hi @AshokDaparthi
> Two questions.
> 
>     1. What's the difference between this one and the commit we approved? Is it just rebased to the latest master commit?
> 
>     2. To what branch will this feature be merged?
>        I was wondering whether we will have a dedicated sonic-mgmt branch for testing sonic switch without this feature. Check [this PR](https://github.com/Azure/sonic-mgmt/pull/3824#issuecomment-906797660) for details
>        @neethajohn FYI.
> 
> 
> Thanks

1# No changes, I did rebase all my pull requests to latest master 
2# Yes, We required sonic-mgmt branch separation before and after merge this changes.

**Review by stephenxs — APPROVED** [review]

**Comment by AshokDaparthi** [comment]
@neethajohn - Could you please review and approve.

**Review by lguohan — APPROVED** [review]

**Comment by zhangyanzhao** [comment]
@neethajohn  can you please help to review and approve this.

**Pull request merged**